### ISSUE TITLE
Type Erased OCKAny[Tasks, Outcomes] interoperability issues

### ISSUE THREAD

**Issue by dwatson0101** [carekit-apple/CareKit#380]
Hi - thanks for resolving the publisher issue I raised in #360 @erik-apple !  I'm trying to now use the sink return object (outcome, task) to serialize into JSON and share with a data manager for integration with a remote server.  That's raising a few questions for me that I'm looking for guidance on on terms of handling type erased OCKAny variants vs concrete types.  For example, OCKTask vs OCKAnyTask.

- How would you recommend serializing a returned OCKAnyTask into JSON?  There doesn't appear to be any methods or encoders defined and it seems a bit cumbersome to convert back into an OCKTask just to take advantage of Codeable conformance.  Realistically, why would OCKTask be Codable if you cannot actually use the type for anything?  It seems like queries, publishers, etc just return the "type erased" variant of OCKAny*.  

- How can I maintain CarePlan integrity with a task or other item that is tied to a care plan?  In a healthcare setting, a patient care plan is a plan with tasks in it for the patient or caregiver to complete.  In CareKit the implementation of that requirement appears to be more rooted in a local database which isn't available as part of the OCKAnyTask protocol.  Thus... if I want to receive updates about adherence, patient updates, etc I loose the ability to tie a task back to a plan when I serialize and sharing with the server without some wonky workarounds.  The same would go if the plan originated from a healthcare setting (care plan + tasks -> dynamically pushed to device).  

E.g.: 

**OCKAnyTask**

```Swift
public protocol OCKAnyTask {

    /// A user-defined unique identifier, typically human readable.
    var id: String { get }

    /// A title that will be used to represent this care plan to the patient.
    var title: String? { get }

    /// Instructions about how this task should be performed.
    var instructions: String? { get }

    /// If true, completion of this task will be factored into the patient's overall adherence. True by default.
    var impactsAdherence: Bool { get }

    /// A schedule that specifies how often this task occurs.
    var schedule: OCKSchedule { get }

    /// A user-defined group identifer that can be used both for querying and sorting results.
    /// Examples may include: "medications", "exercises", "family", "males", "diabetics", etc.
    var groupIdentifier: String? { get }

    /// An identifier for this patient in a remote store.
    var remoteID: String? { get }

    /// Any array of notes associated with this object.
    var notes: [OCKNote]? { get }

    /// Determines if this task belongs to the given care plan.
    ///
    /// - Parameter plan: A care plan that may or may not own this task.
    func belongs(to plan: OCKAnyCarePlan) -> Bool
}
```

**OCKTask**

```Swift
public struct OCKTask: Codable, Equatable, Identifiable, OCKAnyVersionableTask, OCKAnyMutableTask, OCKVersionedObjectCompatible {

    /// The version ID in the local database of the care plan to which this task belongs.
    public var carePlanID: OCKLocalVersionID?

    // MARK: OCKAnyTask
    public let id: String
    public var title: String?
    public var instructions: String?
    public var impactsAdherence = true
    public var schedule: OCKSchedule
    public var groupIdentifier: String?
    public var tags: [String]?
.....
```

**Comment by erik-apple** [comment]
Let's split this answer into two parts!

### What's up with `OCKAnyTask`?
> How would you recommend serializing a returned OCKAnyTask into JSON?

**TLDR;** Cast to `OCKTask` and serialize using standard Codable practices.

Your observation that the publishers produce not-very-useful `OCKAnyTask`s is correct. That was a conscious design choice made for a very specific reason. We don't want to force developers to use `OCKStore` or box them into our notions of what a patient, care plan, task, etc.

The canonical example we always use is somebody who wants to make a CareKit app for veterinarians. Perhaps their notion a "patient" is a `Pet` type that has a `species` field. The store protocol allows them to set their own `Pet` type as the store's patient type. We do require some basic assumptions about the `Pet` type to be true though. For example, that it has a `name` property. If we don't enforce any conventions, then it wouldn't be possible to do things like builds view controllers that show the patient's name.

That minimum set of functionality that we require is codified as `OCKAnyPatient`. Our goal is to keep the type erased protocols well balanced -- minimal enough to be generalizable to almost any use case, but specific enough that we can build meaningful views based the information in them.

**CareKit cannot know what store will be connected or guarantee what types will be used**. However, as the developer of an app, **you can**. If you know that your app is using `OCKStore`, then it's perfectly safe for you to cast from `OCKAnyTask` to `OCKTask` because you have personally guaranteed that only known CareKit types will get pumped through the publisher. 

If you are writing a framework component, it gets a bit trickier. You have to consider the fact that somebody might not use CareKit's native types, and you'll need to have a policy in place for when that happens. In your case, it sounds like your serialization to and from the server is closely tied to `OCKStore`, so you could probably add a `compactMap { $0.task as? OCKTask }` operator onto the task notification publisher to simply ignore data coming from incompatible stores.

### Linking Entities Remotely
You've made a good point. There isn't presently a good way to associate a task with care plan __off device__. This is something we'll have to look into and develop a framework level solution for. As someone working on this problem right now, do you have any insights on how you think the API should be shaped?

**Comment by dwatson0101** [comment]
Thanks for the detailed response.  I assumed casting would be the easiest due to the protocol conformance but wanted to see if there were other thoughts around this.  The context you provided is super helpful!

In regards to number 2 (Linking entities remotely), I'm not entirely sure without fully understanding the impacts around the rest of the framework as I really haven't dug into the UI or much else other than some of the core I/O components.  If I had to make an attempt to answer your question... 

It appears that OCKTask is using the localDataBaseID as part of the OCKVersionable conformance.  My assumption is this was designed so that a query/predicate could be used to stich everything together based on OCKLocalVersionID which must work well for CoreData.  

If I had to guess, I think OCKAnyCarePlan would need an array of OCKAnyTasks to probably match the schema design here: [https://user-images.githubusercontent.com/51756298/69107431-32784800-0a26-11ea-83fc-8987d7ef2e15.png)
](url) 

I'm not sure what that means downstream as I haven't fully absorbed the UI, querying, etc components.  I will also add that I'm a nights and weekend developer and a full time Healthcare guy so I'm approaching it from a different lens and definitely still learning.  




**OCKTask**
...
    // MARK: OCKVersionable
    public var effectiveDate: Date
    public internal(set) var deletedDate: Date?
    public internal(set) var localDatabaseID: OCKLocalVersionID?
    public internal(set) var nextVersionID: OCKLocalVersionID?
    public internal(set) var previousVersionID: OCKLocalVersionID?
...


    public init(id: String, title: String?, carePlanID: OCKLocalVersionID?, schedule: OCKSchedule) {
        self.id = id
        self.title = title
        self.carePlanID = carePlanID
        self.schedule = schedule
        self.effectiveDate = schedule.startDate()
        self.timezone = TimeZone.current
    }
***

**Comment by erik-apple** [comment]
We've recently made a bunch of changes as part of #426 that should make what you're trying to do much easier!

All entities now have a `uuid` property, and entities are associated by using that `uuid` property now. This will make associating entities locally and remotely possible. Please see the linked PR for more details.
